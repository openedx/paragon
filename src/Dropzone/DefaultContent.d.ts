import React from 'react';

export interface DefaultContentProps {
    accept?: { [key: string]: string[] };
    maxSize?: number;
    minSize?: number;
}

declare const DefaultContent: React.FC<DefaultContentProps>;

export default DefaultContent;
