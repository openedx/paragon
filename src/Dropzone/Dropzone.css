/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:20 GMT
 */

:root {
  --pgn-spacing-dropzone-border-error: 2px;
  --pgn-spacing-dropzone-border-active: 2px;
  --pgn-spacing-dropzone-border-focus: 2px;
  --pgn-spacing-dropzone-border-hover: 2px;
  --pgn-spacing-dropzone-border-base: 1px;
  --pgn-spacing-dropzone-padding: 1.5rem;
  --pgn-typography-dropzone-restriction-msg-font-size: var(--pgn-typography-font-size-small-x);
}
