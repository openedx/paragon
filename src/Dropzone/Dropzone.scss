@import "variables";

.pgn__dropzone {
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 200px;
  padding: $dropzone-padding;
  border: $dropzone-border-default;
  box-sizing: border-box;
  cursor: pointer;

  &:hover {
    border: $dropzone-border-hover;
  }

  &:focus {
    border: $dropzone-border-focus;
  }

  &.pgn__dropzone-validation-error {
    border: $dropzone-border-error;
  }

  &.pgn__dropzone-active {
    border: $dropzone-border-active;
  }
}

.pgn__dropzone-error-wrapper {
  text-align: center;
  color: $dropzone-error-wrapper-color;
}

.pgn__dropzone-generic-alert {
  width: 100%;
  margin-bottom: map_get($spacers, 4);

  .pgn__alert-message-wrapper .alert-message-content {
    > span {
      display: block;
    }
  }
}

.pgn__dropzone__upload-icon-wrapper {
  height: 56px;
  width: 56px;
  background: $light-300;
  border-radius: 50%;
  position: relative;
  margin-bottom: map-get($spacers, 4);
}

.pgn__dropzone__upload-icon {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  height: 32px;
  width: 32px;
}

.pgn__dropzone__upload-restriction-message {
  font-size: $dropzone-restriction-msg-font-size;
  color: $dropzone-restriction-msg-color;
}
