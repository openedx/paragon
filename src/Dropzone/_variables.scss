$dropzone-padding:                          1.5rem !default;
$dropzone-border-default:                   1px dashed $gray-500 !default;
$dropzone-border-hover:                     2px solid $info-300 !default;
$dropzone-border-focus:                     2px solid $info-300 !default;
$dropzone-border-active:                    2px solid $primary-500 !default;
$dropzone-border-error:                     2px solid $danger-300 !default;
$dropzone-error-wrapper-color:              $danger-500 !default;
$dropzone-restriction-msg-font-size:        $x-small-font-size !default;
$dropzone-restriction-msg-color:            $gray-500 !default;
