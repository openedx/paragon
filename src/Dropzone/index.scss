:root {
  --pgn-elevation-dropzone-hover:
    var(--pgn-elevation-dropzone-hover-inset)
    var(--pgn-elevation-dropzone-hover-offset-x)
    var(--pgn-elevation-dropzone-hover-offset-y)
    var(--pgn-elevation-dropzone-hover-blur)
    var(--pgn-elevation-dropzone-hover-spread)
    var(--pgn-elevation-dropzone-hover-color);

  --pgn-elevation-dropzone-focus:
    var(--pgn-elevation-dropzone-focus-inset)
    var(--pgn-elevation-dropzone-focus-offset-x)
    var(--pgn-elevation-dropzone-focus-offset-y)
    var(--pgn-elevation-dropzone-focus-blur)
    var(--pgn-elevation-dropzone-focus-spread)
    var(--pgn-elevation-dropzone-focus-color);

  --pgn-elevation-dropzone-error:
    var(--pgn-elevation-dropzone-error-inset)
    var(--pgn-elevation-dropzone-error-offset-x)
    var(--pgn-elevation-dropzone-error-offset-y)
    var(--pgn-elevation-dropzone-error-blur)
    var(--pgn-elevation-dropzone-error-spread)
    var(--pgn-elevation-dropzone-error-color);

  --pgn-elevation-dropzone-active:
    var(--pgn-elevation-dropzone-active-inset)
    var(--pgn-elevation-dropzone-active-offset-x)
    var(--pgn-elevation-dropzone-active-offset-y)
    var(--pgn-elevation-dropzone-active-blur)
    var(--pgn-elevation-dropzone-active-spread)
    var(--pgn-elevation-dropzone-active-color);
}

.pgn__dropzone {
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 200px;
  padding: var(--pgn-spacing-dropzone-padding);
  border: var(--pgn-spacing-dropzone-border-base) dashed var(--pgn-color-dropzone-border-base);
  box-sizing: border-box;
  cursor: pointer;

  &:hover,
  &:focus,
  &.pgn__dropzone-validation-error,
  &.pgn__dropzone-active {
    border-color: transparent;
  }

  &:hover {
    box-shadow: var(--pgn-elevation-dropzone-hover);
  }

  &:focus {
    box-shadow: var(--pgn-elevation-dropzone-focus);
  }

  &.pgn__dropzone-validation-error {
    box-shadow: var(--pgn-elevation-dropzone-error);
  }

  &.pgn__dropzone-active {
    box-shadow: var(--pgn-elevation-dropzone-active);
  }
}

.pgn__dropzone-error-wrapper {
  text-align: center;
  color: var(--pgn-color-dropzone-error-wrapper);
}

.pgn__dropzone-generic-alert {
  width: 100%;
  margin-bottom: var(--pgn-spacing-spacer-4);

  .pgn__alert-message-wrapper .alert-message-content {
    > span {
      display: block;
    }
  }
}

.pgn__dropzone__upload-icon-wrapper {
  height: 56px;
  width: 56px;
  background: var(--pgn-color-light-300);
  border-radius: 50%;
  position: relative;
  margin-bottom: var(--pgn-spacing-spacer-4);
}

.pgn__dropzone__upload-icon {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  height: 32px;
  width: 32px;
}

.pgn__dropzone__upload-restriction-message {
  font-size: var(--pgn-typography-dropzone-restriction-msg-font-size);
  color: var(--pgn-color-dropzone-restriction-msg);
}
