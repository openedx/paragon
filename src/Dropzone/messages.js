import { defineMessages } from 'react-intl';

const messages = defineMessages({
  uploadError: {
    id: 'dropzone.Dropzone.uploadError',
    defaultMessage: 'A problem occured while uploading your file. Please try again.',
    description: 'A message shown in case file upload in Dropzone component results in an error.',
  },
  multipleDragged: {
    id: 'dropzone.Dropzone.multipleDraggedError',
    defaultMessage: 'Only one upload permitted.',
    description: 'A message shown when multiple files are dragged over Dropzone.',
  },
  invalidSizeLess: {
    id: 'dropzone.Dropzone.invalidSizeLessError',
    defaultMessage: 'File must be larger than {size}.',
    description: 'A message shown when a file with less than minimum allowed size is being uploaded in Dropzone.',
  },
  invalidSizeMore: {
    id: 'dropzone.Dropzone.invalidSizeMoreError',
    defaultMessage: 'File must be less than {size}.',
    description: 'A message shown when a file with more than maximum allowed size is being uploaded in Dropzone.',
  },
  invalidType: {
    id: 'dropzone.Dropzone.invalidType',
    defaultMessage: 'The file type must be {count, plural, one {{typeString} file} other {one of {typeString} files}}.',
    description: 'A message shown when a file with wrong MIME type is being uploaded.',
  },
  unexpectedValidationError: {
    id: 'dropzone.Dropzone.unexpectedValidationError',
    defaultMessage: 'An unexpected problem occured during file validation. Please try again.',
    description: 'A message shown in case file validation in Dropzone component for unknown reason.',
  },
  fileSizeBetween: {
    id: 'pgn.Dropzone.DefaultContent.fileSizeBetween',
    defaultMessage: 'Between {sizeMin} and {sizeMax}',
    description: "A message shown when uploaded file's size must be in given range.",
  },
  fileSizeMax: {
    id: 'pgn.Dropzone.DefaultContent.fileSizeMax',
    defaultMessage: 'Max {sizeMax}',
    description: "A message shown when uploaded file's size must be more than some value.",
  },
  fileSizeMin: {
    id: 'pgn.Dropzone.DefaultContent.fileSizeMin',
    defaultMessage: 'Min {sizeMin}',
    description: "A message shown when uploaded file's size must be more than some value.",
  },
  fileTypeRestriction: {
    id: 'pgn.Dropzone.DefaultContent.fileTypeRestriction',
    defaultMessage: 'Upload {count, plural, one {{firstPart} files} other {{firstPart} or {secondPart} files}}',
    description: 'A message shown when uploaded file must be of certain type(s).',
  },
});

export default messages;
