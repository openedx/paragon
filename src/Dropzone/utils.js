/**
 * Checks whether accept object contains multiple MIME types or
 * file extensions (or their combination).
 * @param {object} accept - An Object with the keys set to the MIME type and the values an Array of file extensions.
 */
const isMultipleTypes = (accept) => {
  if (Object.keys(accept).length > 1) {
    return true;
  }
  if (Object.keys(accept).length === 0) {
    return false;
  }
  return Object.values(accept)[0].length > 1;
};

/**
 * Returns string representation of file extensions from 'accept' object.
 * @param {object} accept - An Object with the keys set to the MIME type and the values an Array of file extensions
 */
const getTypesString = (accept) => Object.entries(accept).reduce((previousValue, currentValue) => {
  const [mimeType, fileExtensions] = currentValue;
  let value;

  if (fileExtensions.length > 0) {
    value = `${fileExtensions.join(', ').replaceAll('.', '').toUpperCase()}, `;
  } else if (mimeType.endsWith('/*')) {
    value = `${mimeType.slice(0, -2)}, `;
  } else {
    value = `${mimeType.split('/').pop().toUpperCase()}, `;
  }

  return previousValue + value;
}, '').slice(0, -2);

/**
 * Returns human-readable form of bytes.
 * @param {Number} bytes - value in bytes.
 */
const formatBytes = (bytes) => {
  if (bytes === 0) {
    return '0 Bytes';
  }

  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'];

  const i = Math.floor(Math.log(bytes) / Math.log(k));

  return `${parseFloat((bytes / k ** i).toFixed(2))}${sizes[i]}`;
};

export { getTypesString, isMultipleTypes, formatBytes };
