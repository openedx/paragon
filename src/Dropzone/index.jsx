import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { useDropzone, ErrorCode } from 'react-dropzone';
import { fromEvent } from 'file-selector';
import { useIntl } from 'react-intl';

import DragError from './DragError';
import GenericError from './GenericError';
import UploadProgress from './UploadProgress';
import DefaultContent from './DefaultContent';
import messages from './messages';
import { getTypesString, isMultipleTypes, formatBytes } from './utils';

function Dropzone({
  className, accept, minSize, maxSize, validator,
  errorMessages, progressVariant, inputComponent,
  onProcessUpload, onUploadProgress, onUploadCancel,
  ...props
}) {
  const [isMultipleDragged, setIsMultipleDragged] = useState(false);
  const [errors, setErrors] = useState([]);
  const [progress, setProgress] = useState(0);
  const [fileName, setFileName] = useState(undefined);
  const [controller, setController] = useState(undefined);
  const intl = useIntl();

  const {
    uploadError: uploadErrorMsg,
    invalidSizeLess: invalidSizeLessMsg,
    invalidSizeMore: invalidSizeMoreMsg,
    invalidType: invalidTypeMsg,
    multipleDragged: multipleDraggedMsg,
  } = errorMessages;

  const onDragEnter = async (e) => {
    if (errors) {
      setErrors([]);
    }
    const files = await fromEvent(e);
    if (files && files.length > 1) {
      setIsMultipleDragged(true);
    }
  };

  const onDragLeave = () => {
    if (isMultipleDragged) {
      setIsMultipleDragged(false);
    }
  };

  const onDropRejected = (files) => {
    if (!isMultipleDragged) {
      setErrors(files[0].errors.map(error => {
        switch (error.code) {
          case ErrorCode.FileTooLarge:
            return invalidSizeMoreMsg || intl.formatMessage(messages.invalidSizeMore, { size: formatBytes(maxSize) });
          case ErrorCode.FileTooSmall:
            return invalidSizeLessMsg || intl.formatMessage(messages.invalidSizeLess, { size: formatBytes(minSize) });
          case ErrorCode.FileInvalidType:
            return invalidTypeMsg || intl.formatMessage(
              messages.invalidType,
              { count: isMultipleTypes(accept) ? 2 : 1, typeString: getTypesString(accept) },
            );
          default:
            return intl.formatMessage(messages.unexpectedValidationError);
        }
      }));
    } else {
      setIsMultipleDragged(false);
    }
  };

  const handleProgressUpload = (progressEvent) => {
    const percentValue = Math.round((progressEvent.loaded * 100) / progressEvent.total);
    setProgress(percentValue);
    onUploadProgress(percentValue, progressEvent);
  };

  const handleUploadError = (error) => {
    // check if request has been canceled before treating the exception as an upload error
    if (error.code !== 'ERR_CANCELED') {
      setProgress(0);
      setErrors([uploadErrorMsg || intl.formatMessage(messages.uploadError)]);
    }
  };

  const processUpload = (fileData) => {
    const newController = new AbortController();
    setController(newController);

    const requestConfig = {
      onUploadProgress: handleProgressUpload,
      signal: newController.signal,
    };

    onProcessUpload({
      fileData,
      requestConfig,
      handleError: handleUploadError,
    });
  };

  const onDropAccepted = async (files) => {
    const file = files[0];
    if (validator) {
      const customValidationError = await validator(file);
      if (customValidationError) {
        setErrors([customValidationError]);
        return;
      }
    }

    if (errors) {
      setErrors([]);
    }

    const formData = new FormData();
    formData.append('file', file);
    setFileName(file.name);

    processUpload(formData);
  };

  const handleUploadCancel = () => {
    controller.abort();
    setProgress(0);
    onUploadCancel();
  };

  const {
    getRootProps,
    getInputProps,
    isDragActive,
    isDragReject,
  } = useDropzone({
    multiple: false,
    maxFiles: 1,
    maxSize,
    minSize,
    onDragLeave,
    onDragEnter,
    onDropRejected,
    onDropAccepted,
    accept,
    disabled: progress && progress !== 100,
  });

  const renderContent = () => {
    if (isMultipleDragged) {
      return <DragError message={multipleDraggedMsg || intl.formatMessage(messages.multipleDragged)} />;
    }

    if (errors.length > 0) {
      return (
        <>
          <GenericError errorMsgs={errors} />
          {inputComponent || <DefaultContent minSize={minSize} maxSize={maxSize} accept={accept} />}
        </>
      );
    }

    if (progress && progress !== 100) {
      return (
        <UploadProgress
          variant={progressVariant}
          percent={progress}
          name={fileName}
          onCancel={handleUploadCancel}
        />
      );
    }

    return inputComponent || <DefaultContent minSize={minSize} maxSize={maxSize} accept={accept} />;
  };

  return (
    <div
      data-testid="dropzone-container"
      {...getRootProps({
        className: classNames('pgn__dropzone', className, {
          'pgn__dropzone-validation-error': isMultipleDragged || errors.length > 0 || isDragReject,
          'pgn__dropzone-active': isDragActive && !isDragReject,
        }),
      })}
      {...props}
    >
      <input {...getInputProps()} />
      <div className="d-flex flex-column justify-content-around align-items-center w-100">
        {renderContent()}
      </div>
    </div>
  );
}

Dropzone.defaultProps = {
  className: undefined,
  accept: undefined,
  maxSize: Infinity,
  minSize: 0,
  onUploadProgress: () => {},
  onUploadCancel: () => {},
  errorMessages: {
    invalidType: undefined,
    invalidSizeLess: undefined,
    invalidSizeMore: undefined,
    multipleDragged: undefined,
    uploadError: undefined,
  },
  progressVariant: 'spinner',
  validator: undefined,
  inputComponent: undefined,
};

Dropzone.propTypes = {
  /** Specifies class name to append to the base element. */
  className: PropTypes.string,
  /**
   * Set accepted file types.
   * This should be an object with the keys set to the
   * [MIME type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types)
   * and the values to an array of file extensions.
   */
  accept: PropTypes.objectOf(PropTypes.arrayOf(PropTypes.string)),
  /** Maximum file size (in bytes). */
  maxSize: PropTypes.number,
  /** Minimum file size (in bytes). */
  minSize: PropTypes.number,
  /**
   * A callback fired each time an upload progress event happens,
   * receives (percentageUploaded, progressEvent) as arguments.
   */
  onUploadProgress: PropTypes.func,
  /** A callback fired upon successful upload, receives Response object as a single argument. */
  onUploadCancel: PropTypes.func,
  /**
   * A function responsible for uploading the file.
   * Receives following object as its only argument
   * {
   *   @param {object} fileData - Metadata about the uploaded file.
   *   @param {object} requestConfig - Config to pass to `axios` call.
   *   @param {function} handleError - Function to communicate to `Dropzone` that file upload resulted in failure,
   *   expects `Error` object to be passed as its only argument.
   * }
   */
  onProcessUpload: PropTypes.func.isRequired,
  /**
   * An object containing error messages, following are supported:
   * 1) invalidType - A message to display when file of invalid type is dropped into `Dropzone`.
   * Defaults to 'The file type must be {filType} file / one of {fileTypes} files.'.
   * 2) invalidSizeLess - A message to display when file of size less than minSize value is dropped into `Dropzone`.
   * Defaults to 'File must be larger than {minSize}.'.
   * 3) invalidSizeMore - A message to display when file of size greater than maxSize value is dropped into `Dropzone`.
   * Defaults to 'File must be less than {maxSize}.'.
   * 4) multipleDragged - A message to display when multiple files are dragged over `Dropzone`.
   * 5) uploadError - A message to display in case upload results in an error
   */
  errorMessages: PropTypes.shape({
    invalidType: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    invalidSizeLess: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    invalidSizeMore: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    multipleDragged: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    uploadError: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
  }),
  /** Specifies how the upload progress should be displayed, component shows either spinner or a progress bar. */
  progressVariant: PropTypes.oneOf(['spinner', 'bar']),
  /**
   * Custom validation function, receives `File` object as its only argument.
   * Note that this function will be invoked as a last validation step before beginning an upload process.
   */
  validator: PropTypes.func,
  /** A component to display initial state of the `Dropzone`. */
  inputComponent: PropTypes.oneOfType([PropTypes.elementType, PropTypes.node]),
};

export default Dropzone;
