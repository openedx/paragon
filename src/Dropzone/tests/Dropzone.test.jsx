import React from 'react';
import { IntlProvider } from 'react-intl';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import renderer from 'react-test-renderer';
import {
  render,
  act,
  fireEvent,
  screen,
} from '@testing-library/react';
import '@testing-library/jest-dom';

import { formatBytes } from '../utils';
import messages from '../messages';
import Dropzone from '..';

const axiosMock = new MockAdapter(axios);
axiosMock.onPost('/upload').reply((config) => {
  const total = 1024; // mocked file size
  const progress = 0.4;
  if (config.onUploadProgress) {
    config.onUploadProgress({ loaded: total * progress, total });
  }
  return new Promise(() => {});
});

// dummy function to test that upload event was not fired (i.e. this function has never been called)
// if your test case tests successful upload action you should override this function, see examples below
const handleProcessUpload = jest.fn();

function createFile(name, size, type) {
  const file = new File([], name, { type });
  Object.defineProperty(file, 'size', {
    get() {
      return size;
    },
  });
  return file;
}

function createDtWithFiles(files = []) {
  return {
    dataTransfer: {
      files,
      items: files.map((file) => ({
        kind: 'file',
        size: file.size,
        type: file.type,
        getAsFile: () => file,
      })),
      types: ['Files'],
    },
  };
}

const imageFile = createFile('image.png', 111, 'image/png');
const pdfFile = createFile('something.pdf', 10000, 'application/pdf');

// eslint-disable-next-line react/prop-types
function DropzoneWrapper({ children, ...props }) {
  return (
    <IntlProvider locale="en" messages={{}}>
      <Dropzone onProcessUpload={handleProcessUpload} {...props} />
    </IntlProvider>
  );
}

describe('<Dropzone />', () => {
  it('successfully renders', () => {
    const tree = renderer.create(<DropzoneWrapper />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('call onProcessUpload func if file passes validation', async () => {
    const processUpload = jest.fn();
    render(<DropzoneWrapper onProcessUpload={processUpload} />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(processUpload).toHaveBeenCalledTimes(1);
  });

  it('renders warning if multiple files are dragged and does nothing if files are dropped', async () => {
    render(<DropzoneWrapper />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile, pdfFile]);
    await act(async () => {
      fireEvent.dragEnter(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent('Only one upload permitted.');
    await act(async () => {
      fireEvent.dragLeave(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent('Drag and drop your file here or click to upload.');
    await act(async () => {
      fireEvent.dragEnter(dropzoneContainer, dt);
    });
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(handleProcessUpload).toHaveBeenCalledTimes(0);
    expect(dropzoneContainer).toHaveTextContent('Drag and drop your file here or click to upload.');
  });

  it('does not upload files that exceed maxSize and renders default error message', async () => {
    render(<DropzoneWrapper maxSize={1000} />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([pdfFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent(messages.invalidSizeMore.defaultMessage.replace('{size}', formatBytes(1000)));
    expect(handleProcessUpload).toHaveBeenCalledTimes(0);
  });

  it('does not upload files that have less size than minSize and renders default error message', async () => {
    render(<DropzoneWrapper minSize={1000} />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent(messages.invalidSizeLess.defaultMessage.replace('{size}', formatBytes(1000)));
    expect(handleProcessUpload).toHaveBeenCalledTimes(0);
  });

  it('does not upload files that have wrong extension and renders default error message', async () => {
    render(<DropzoneWrapper accept={{ 'application/pdf': [] }} />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent('The file type must be PDF file.');
    expect(handleProcessUpload).toHaveBeenCalledTimes(0);
  });

  it('renders multiple error messages if there multiple validation failures', async () => {
    render(<DropzoneWrapper maxSize={100} accept={{ 'application/pdf': [] }} />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent('The file type must be PDF file.');
    expect(dropzoneContainer).toHaveTextContent(messages.invalidSizeMore.defaultMessage.replace('{size}', formatBytes(100)));
    expect(handleProcessUpload).toHaveBeenCalledTimes(0);
  });

  it('renders custom error messages', async () => {
    render(
      <DropzoneWrapper
        maxSize={1000}
        minSize={150}
        accept={{ 'application/pdf': [] }}
        errorMessages={{
          invalidType: 'Invalid type error msg.',
          invalidSizeLess: 'Size less error msg.',
          invalidSizeMore: 'Size more error msg.',
        }}
      />,
    );
    const dropzoneContainer = await screen.findByTestId('dropzone-container');

    const dtImage = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dtImage);
    });
    expect(dropzoneContainer).toHaveTextContent('Size less error msg.');
    expect(dropzoneContainer).toHaveTextContent('Invalid type error msg.');

    const dtPdf = createDtWithFiles([pdfFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dtPdf);
    });
    expect(dropzoneContainer).toHaveTextContent('Size more error msg.');
    expect(handleProcessUpload).toHaveBeenCalledTimes(0);
  });

  it('renders custom validation error message returned by validator func', async () => {
    const validator = () => 'Custom validation error message.';
    render(<DropzoneWrapper validator={validator} />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent('Custom validation error message.');
  });

  it('renders progress component when upload progress is not 0', async () => {
    const handleProcessUploadMock = async ({
      fileData, requestConfig, handleError,
    }) => {
      const uploadUrl = '/upload';
      try {
        await axios.post(uploadUrl, fileData, requestConfig);
      } catch (error) {
        handleError(error);
      }
    };

    render(<DropzoneWrapper onProcessUpload={handleProcessUploadMock} />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent('Uploading image.png, 40% done.');
  });

  it('resets default Dropzone state after pressing cancel button', async () => {
    const handleProcessUploadMock = async ({
      fileData, requestConfig, handleError,
    }) => {
      const uploadUrl = '/upload';
      try {
        await axios.post(uploadUrl, fileData, requestConfig);
      } catch (error) {
        handleError(error);
      }
    };

    render(<DropzoneWrapper onProcessUpload={handleProcessUploadMock} progressVariant="bar" />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent('Uploading image.png.');

    const cancelButton = screen.getByRole('button', { name: 'Cancel' });
    fireEvent.click(cancelButton);
    expect(dropzoneContainer).toHaveTextContent('Drag and drop your file here or click to upload.');
  });

  it('shows upload error message in case upload failed', async () => {
    const handleProcessUploadWithError = async ({ handleError }) => handleError({});
    render(<DropzoneWrapper onProcessUpload={handleProcessUploadWithError} progressVariant="bar" />);
    const dropzoneContainer = await screen.findByTestId('dropzone-container');
    const dt = createDtWithFiles([imageFile]);
    await act(async () => {
      fireEvent.drop(dropzoneContainer, dt);
    });
    expect(dropzoneContainer).toHaveTextContent(messages.uploadError.defaultMessage);
  });
});
