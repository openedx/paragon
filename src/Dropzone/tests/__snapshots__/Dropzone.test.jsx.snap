// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropzone /> successfully renders 1`] = `
<div
  className="pgn__dropzone"
  data-testid="dropzone-container"
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  role="presentation"
  tabIndex={0}
>
  <input
    multiple={false}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
  <div
    className="d-flex flex-column justify-content-around align-items-center w-100"
  >
    <div
      className="pgn__dropzone__upload-icon-wrapper"
    >
      <span
        className="pgn__icon pgn__dropzone__upload-icon"
      >
        <svg
          aria-hidden={true}
          fill="none"
          focusable={false}
          height={24}
          role="img"
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 10h4v6h6v-6h4l-7-7-7 7zm0 8v2h14v-2H5z"
            fill="currentColor"
          />
        </svg>
      </span>
    </div>
    <p>
      Drag and drop your file here or click to upload.
    </p>
    <p
      className="pgn__dropzone__upload-restriction-message"
    />
  </div>
</div>
`;
